src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -0,0 +1,177 @@
+const validWordSymbols = "\\d._'$@&/";
+const modifierSymbols = '\\!+-';
+
+// Basic patterns
+const word = `[${letterSymbols}${validWordSymbols}]{1}[${letterSymbols}${validWordSymbols}${modifierSymbols}]*`;
+const wordModifier = `[${modifierSymbols}]{1}`;
+const separator = `[${spaceSymbols}]+`;
+const separatorOrEnd = `(?:${separator}|$)`;
+const parentheses = '(?:\\(|\\)|\\[|\\]|\\{|\\}|\\<|\\>)';
+const wildcard = '\\*';
+const phrase = '(?:"[^"]*"|\'[^\']*\')';
+const simpleOperator = '(?:OR|AND|NEAR|NOT|\\&|\\!|\\|)';
+const andNotOperator = `AND${separator}NOT`;
+const aroundOperator = `AROUND${separator}\\d+`;
+
+export class BooleanQueryTokenizer {
+    /**
+     * Token patterns are tuples that define a token type and its pattern.
+     * Token patterns are ordered from low to high priority.
+     * Token priorities are used as a tie breakers when multiple patterns matched
+     * the same lexeme. E.g. lexeme "NEAR" is a boolean operator and a word at the same
+     * time; it will be recognized as a `booleanOperator` as long as `booleanOperator` has
+     * higher priority than `word`.
+     * Some context specific patterns use lookahead assertions: /^pattern(?=lookahead)/
+     */
+    private static readonly tokenPatterns: TokenPattern[] = [
+        [TokenType.separator, new RegExp(`^(?:${separator}|$)`)],
+        [TokenType.word, new RegExp(`^${word}`)],
+        [TokenType.modifier, new RegExp(`^${wordModifier}(?=${word})`)],
+        [
+            TokenType.wildcard,
+            new RegExp(`^(?:${wildcard}(?=${word})|${wildcard}(?=${separatorOrEnd}))`),
+        ],
+        [TokenType.wildcardWord, new RegExp(`^${word}${wildcard}${word}`)],
+        [TokenType.phrase, new RegExp(`^${phrase}`)],
+        [
+            TokenType.booleanOperator,
+            new RegExp(
+                `^(?:${andNotOperator}|${aroundOperator}|${simpleOperator})(?=${separatorOrEnd})`
+            ),
+        ],
+        [TokenType.parentheses, new RegExp(`^${parentheses}`)],
+    ];
+
+    /**
+     * Tokenizer splits an input text into meaningful chunks and labels each chunk according to its
+     * lexical meaning, suchwise producing a list of tokens.
+     */
+    // eslint-disable-next-line class-methods-use-this
```

Comment:
Why not make all functions static. That makes sense, there is no actual need to initialise the class or have different instances of it running somewhere. It can be used simply like a little library.

Reply:
Makes sense, updated